cpp/arcticdb/storage/azure/azure_real_client.cpp

```diff
@@ -0,0 +1,115 @@
+/* Copyright 2023 Man Group Operations Limited
+ *
+ * Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
+ *
+ * As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+ */
+
+
+#include <azure/core/http/curl_transport.hpp>
+#include <azure/core.hpp>
+#include <azure/storage/blobs.hpp>
+
+#include <arcticdb/storage/storage_utils.hpp>
+#include <arcticdb/storage/azure/azure_real_client.hpp>
+#include <arcticdb/storage/azure/azure_client_wrapper.hpp>
+#include <arcticdb/storage/object_store_utils.hpp>
+
+namespace arcticdb::storage {
+using namespace object_store_utils;
+namespace azure {
+using namespace Azure::Storage;
+using namespace Azure::Storage::Blobs;
+
+Azure::Core::Context get_context(unsigned int request_timeout){
+    Azure::Core::Context requestContext; //TODO: Maybe can be static but need to be careful with its shared_ptr and ContextSharedState
+    return requestContext.WithDeadline(std::chrono::system_clock::now() + std::chrono::milliseconds(request_timeout));
+}
+
+
+RealAzureClient::RealAzureClient(const Config &conf) :
+container_client(BlobContainerClient::CreateFromConnectionString(conf.endpoint(), conf.container_name(), get_client_options(conf))) { }
+
+Azure::Storage::Blobs::BlobClientOptions RealAzureClient::get_client_options(const Config &conf) {
+    BlobClientOptions client_options;
+    if (!conf.ca_cert_path().empty()) {//WARNING: Setting ca_cert_path will force Azure sdk uses libcurl as backend support, instead of winhttp
+        Azure::Core::Http::CurlTransportOptions curl_transport_options;
+        curl_transport_options.CAInfo = conf.ca_cert_path();
+        client_options.Transport.Transport = std::make_shared<Azure::Core::Http::CurlTransport>(curl_transport_options);
+    }
+    return client_options;
+}
+
+void RealAzureClient::write_blob(
+        const std::string& blob_name,
+        Segment&& segment,
+        const Azure::Storage::Blobs::UploadBlockBlobFromOptions& upload_option,
+        unsigned int request_timeout) {
+
+    std::shared_ptr<Buffer> tmp;
+    auto hdr_size = segment.segment_header_bytes_size();
+    auto [dst, write_size] = segment.try_internal_write(tmp, hdr_size);
+    util::check(arcticdb::Segment::FIXED_HEADER_SIZE + hdr_size + segment.buffer().bytes() <= write_size,
+                "Size disparity, fixed header size {} + variable header size {} + buffer size {}  >= total size {}",
+                arcticdb::Segment::FIXED_HEADER_SIZE,
+                hdr_size,
+                segment.buffer().bytes(),
+                write_size);
+    ARCTICDB_SUBSAMPLE(AzureStorageUploadObject, 0)
```

Comment:
Why is the section above here (calculating dst and write size) in the real vs mock client API? I would think we want to execute the segment logic above regardless of whether we are using mock or real azure client?

Reply:
Because I need `dst` and `write_size` for `blob_client.UploadFrom` in real client. For the mock client, I don't need `dst` and `write_size` I shall directly save the `segment` in a map. Have a look at Ivo's similar logic #1274.